### PR TITLE
JCache config tests

### PR DESCRIPTION
**Pull request by cherednichenko**
test's for issue #4753

### CODE REVIEW

**Comment by bilalyasar** [comment]
lets run tests again, because schema validation PR has been merged.
verify

**Comment by cherednichenko** [comment]
Should I run those tests again - I didn't catch?

Actually those tests are passing after Serkan fixed #4753 issue - we re-run with Danny before creating this PR.

**Comment by bilalyasar** [comment]
no, after writing [v]erify, jenkins automatically triggers new build.
and seems it passed. :+1:

**Comment by cherednichenko** [comment]
Ok. :+1:

**Comment by pveentjer** [comment]
@serkan-ozal can you have a look at this PR.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                    HazelcastServerCachingProvider cachingProvider = HazelcastServerCachingProvider
+                            .createCachingProvider(hazelcastInstance);
+
+                    Cache<Object, Object> cache = cachingProvider.getCacheManager().getCache("xmlCache" );
+                    cache.get(1);
+                    latch.countDown();
+                }
+            }.start();
+        }
+        HazelcastTestSupport.assertOpenEventually(latch);
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @cherednichenko even if this test fails (I know it shouldn't fail :) ) we should be sure that all threads are terminated.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                    HazelcastServerCachingProvider cachingProvider = HazelcastServerCachingProvider
+                            .createCachingProvider(hazelcastInstance);
+
+                    Cache<Object, Object> cache = cachingProvider.getCacheManager().getCache("xmlCache"+finalI);
+                    cache.get(1);
+                    latch.countDown();
+                }
+            }.start();
+        }
+        HazelcastTestSupport.assertOpenEventually(latch);
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @cherednichenko even if this test fails (I know it shouldn't fail :) ) we should be sure that all threads are terminated.

File: hazelcast/src/test/resources/test-hazelcast-real-jcache.xml [hunk 3]
```diff
@@ -0,0 +1,80 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (serkan-ozal) [inline hazelcast/src/test/resources/test-hazelcast-real-jcache.xml]: Is this `test-hazelcast-real-jcache.xml` used only by this test ? If it is, it may better to clean unnecessary configurations from this xml configuration to make it clean.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+@Category(SlowTest.class)
+public class CacheCreationTest {
+
+    private final URL configUrl1 = getClass().getClassLoader().getResource("test-hazelcast-real-jcache.xml");
+    Config hzConfig;
+
+    @Before
+    public void init() throws Exception{
+        Hazelcast.shutdownAll();
+        XmlConfigBuilder configBuilder = new XmlConfigBuilder(configUrl1.getFile());
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: `hzConfig` may be initialized only once in another method `@BeforeClass` and can be defined static.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    @After
+    public void cleanup() throws Exception{
+        Hazelcast.shutdownAll();
+    }
+
+    @Test
+    public void creatingASingleCache() throws URISyntaxException {
+
+        HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(hzConfig);
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Can't we use test Hazelcast instance via `HazelcastTestSupport` ? Do we really need network stuff for this test ?

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+
+    @Test
+    public void creatingASingleCacheFromMultiProviders() throws URISyntaxException {
+
+        int threadCount=4;
+
+        final CountDownLatch latch = new CountDownLatch(threadCount);
+        for (int i = 0; i < threadCount; i++) {
+            final int finalI = i;
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Seems no need to `finalI` for this test

File: hazelcast/src/test/resources/test-hazelcast-real-jcache.xml [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    </cache>
+
+    <cache name="testCache">
+        <key-type class-name="java.lang.Integer"/>
+        <value-type class-name="java.lang.String"/>
+        <statistics-enabled>false</statistics-enabled>
+        <management-enabled>false</management-enabled>
+    </cache>
+
+    <cache name="wanRefTestCache">
```
Comment (serkan-ozal) [inline hazelcast/src/test/resources/test-hazelcast-real-jcache.xml]: I think, `wanRefTestCache` cache config is not needed if this xml configuration file is not used by other tests.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            new Thread() {
+                public void run() {
+                    HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(hzConfig);
+                    HazelcastServerCachingProvider cachingProvider = HazelcastServerCachingProvider
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Could you make this tests also for client based cache provider ? If you think, it is not scope of this test, could you create another test or test file ?

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Could you make this tests also for client based cache provider ? If you think, it is not scope of this test, could you create another test or test file ?

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    @After
+    public void cleanup() throws Exception{
+        Hazelcast.shutdownAll();
+    }
+
+    @Test
+    public void creatingASingleCache() throws URISyntaxException {
+
+        HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(hzConfig);
+        HazelcastServerCachingProvider cachingProvider = HazelcastServerCachingProvider
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Could you make this tests also for client based cache provider ? If you think, it is not scope of this test, could you create another test or test file ?

**Comment by pveentjer** [comment]
@hasancelik can you help @cherednichenko out to resolve the review comments?

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 1 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @serkan-ozal Could you consult me how it could be done? Because I can't find suitable form for it in our code. 
I quite understand what exactly should be checked BUT your advise is highly appreciated.

Thanks a lot.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 2 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @serkan-ozal Could you consult me how it could be done? Because I can't find suitable form for it in our code. 
I quite understand what exactly should be checked BUT your advise is highly appreciated.

p.s. the same as for previous comment.

File: hazelcast/src/test/resources/test-hazelcast-real-jcache.xml (hunk 3 above)
Comment (cherednichenko) [inline hazelcast/src/test/resources/test-hazelcast-real-jcache.xml]: Seems yes - it uses only for this test as as I remember :) 

Do you mean remove all and keep only this part ? 

```
 <cache name="xmlCache*">
    </cache>
```

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 4 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: I added:

```
@RunWith(HazelcastSerialClassRunner.class)
@Category(SlowTest.class)
public class CacheCreationTest extends HazelcastTestSupport {

    static Config hzConfig;

    @BeforeClass
    public void init() throws Exception{
        final URL configUrl1 = getClass().getClassLoader().getResource("test-hazelcast-real-jcache.xml");
        XmlConfigBuilder configBuilder = new XmlConfigBuilder(configUrl1.getFile());
        hzConfig = configBuilder.build();
    }
```

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 5 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Added:

```
@RunWith(HazelcastSerialClassRunner.class)
@Category(SlowTest.class)
public class CacheCreationTest extends HazelcastTestSupport {
   .......
}
```

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 6 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: removed.

File: hazelcast/src/test/resources/test-hazelcast-real-jcache.xml (hunk 7 above)
Comment (cherednichenko) [inline hazelcast/src/test/resources/test-hazelcast-real-jcache.xml]: It was merged I believe when we pushed/committed.

Removed.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 4 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @cherednichenko `@BeforeClass` annotations must be tagged to static methods. So `init` method must be static

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 8 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: I believe that it would be better to create as different commit.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 5 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: :+1:

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 9 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: I believe that it would be better to create as different commit and different test for client.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 10 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: I believe that it would be better to create as different commit and different test for client.

**Comment by cherednichenko** [comment]
@pveentjer @hasancelik helped me and we did all changes except one - related to terminated threads.

I'll finish and commit my changes.

Thanks a lot guys.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 1 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @cherednichenko you may use `java.util.concurrent.ExecutorService` to wait threads to finish with a timeout with `awaitTimeout` method. If they still don't terminate, you may use `shutdown`

File: hazelcast/src/test/resources/test-hazelcast-real-jcache.xml (hunk 3 above)
Comment (serkan-ozal) [inline hazelcast/src/test/resources/test-hazelcast-real-jcache.xml]: yep

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 10 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: ok

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 8 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: ok

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 1 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @serkan-ozal  done.

File: hazelcast/src/test/resources/test-hazelcast-real-jcache.xml (hunk 3 above)
Comment (cherednichenko) [inline hazelcast/src/test/resources/test-hazelcast-real-jcache.xml]: done.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 4 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @serkan-ozal  Yeap - you're right :)
Done.

**Comment by cherednichenko** [comment]
I made all changes based on comments to this PR, did some optimisations, checked/run 'checkstyle' and 'findbugs', run all tests -- everything is GREEN, no issues/failures.

Thanks a lot guys ! :+1:

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+                public void run() {
+                    HazelcastServerCachingProvider cachingProvider = createCachingProvider(hzConfig);
+                    Cache<Object, Object> cache = cachingProvider.getCacheManager().getCache("xmlCache"+finalI);
+                    cache.get(1);
+                    latch.countDown();
+                }
+            });
+        }
+        HazelcastTestSupport.assertOpenEventually(latch);
+        if (!execSer.awaitTermination(3, TimeUnit.SECONDS)) {
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: can we make it `30` seconds. `3` seconds may not be sufficient since it is also created an hazelcast instance

Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Done.

p.s. By the way, this small change fixes issue which I've seen:
```
when I run one test after another (within this class) - there is some situation which is similar to this error occurred before -- "Node factory is already created!". But when I waited +1-2-3 secs more --- all tests are passed.
```

**Comment by serkan-ozal** [comment]
from my side, it is ok. :+1:

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                public void run() {
+                    HazelcastServerCachingProvider cachingProvider = createCachingProvider(hzConfig);
+                    Cache<Object, Object> cache = cachingProvider.getCacheManager().getCache("xmlCache"+finalI);
+                    cache.get(1);
+                    latch.countDown();
+                }
+            });
+        }
+        HazelcastTestSupport.assertOpenEventually(latch);
+        if (!execSer.awaitTermination(30, TimeUnit.SECONDS)) {
```
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: sorry, I forgot to mention that you may also move wait and shutdown codes to a finally block to be sure that even at failure case, these will be shutdown.

Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Just put it in the @Before/@After methods of a test. I also see that you are doing 

assertEquals(0, HazelcastClient.getAllHazelcastClients().size());

This is also highly undesirable. Because it make a test difficult to understand and apparently you are not trusting something and you let this lack of confidence leak in your test. If could very well be that there is an issue; if so.. it needs to be addressed. Otherwise you should only rely on before/after methods to set your tests in a valid state.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    public void creatingASingleCache() throws URISyntaxException {
+        HazelcastServerCachingProvider cachingProvider = createCachingProvider(hzConfig);
+        Cache<Object, Object> cache = cachingProvider.getCacheManager().getCache("xmlCache" + 1);
+        cache.get(1);
+    }
+
+    @Test
+    public void creatingASingleCacheFromMultiProviders() throws URISyntaxException, InterruptedException {
+
+        ExecutorService execSer = Executors.newFixedThreadPool(THREAD_COUNT);
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Please don't use partial names because it makes code very difficult to understand. Just call it executorService or executor.

Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Also don't add constants if they have meaning for only a single test. In this case the number of threads you are using is not relevant for the rest of the system. So just declare a local variable ' int threadcount = 10' or something.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    @Test
+    public void creatingASingleCache() throws URISyntaxException {
+        HazelcastServerCachingProvider cachingProvider = createCachingProvider(hzConfig);
+        Cache<Object, Object> cache = cachingProvider.getCacheManager().getCache("xmlCache" + 1);
+        cache.get(1);
+    }
+
+    @Test
+    public void creatingASingleCacheFromMultiProviders() throws URISyntaxException, InterruptedException {
+
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Don't use empty white lines at the beginning of a method.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(SlowTest.class)
+public class CacheCreationTest extends HazelcastTestSupport {
+
+    static Config hzConfig;
+    private static final int THREAD_COUNT = 4;
+
+    @BeforeClass
+    public static void init() throws Exception{
+        final URL configUrl1 = CacheCreationTest.class.getClassLoader().getResource("test-hazelcast-real-jcache.xml");
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Don't number local variables.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 11 above)
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: This code is not correct. You can't wait for termination and then do a shutdown. 

So delete the if...shudown part and just say shutdown. That is sufficient.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 14 above)
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: The test name is not really revealing what you are doing. I'm confused about why you need multiple threads to do what you are doing. I think this test can be simplified by removing the threads completely.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+        }
+        HazelcastTestSupport.assertOpenEventually(latch);
+        if (!execSer.awaitTermination(3, TimeUnit.SECONDS)) {
+            execSer.shutdown();
+        }
+    }
+
+    @Test
+    public void creatingMultiCacheFromMultiProviders() throws URISyntaxException, InterruptedException {
+
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: No empty lines.

Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: In case of test I just say 'throws Exception' I don't care which exceptions are thrown. They are all going to cause the test to fail.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 13 above)
Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: if a variable's value is not changed, it should be defined as contant. this is common/suggested usecase

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 15 above)
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: I think it would be better of this test does not make use of a XML file to get the configuration from; my experience is that having all these test xml files make the system very unmaintainable. Just make a string to stick the configuration in and if you need to have a url, just create a tmp file.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 13 above)
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: The problem is that when you get a test with a need for different values for the same 'constant', you run into a problems. Now you are forcing everyone to use 10 thread or whatever number you have defined. If you want to add a test that only need 2 threads, then you add ANOTHER_THREAD_COUNT constant?

Comment (serkan-ozal) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: I don't still understand your objection. This constant is defined in this test method's scope. Every test methods can use their constants in themselves. There is no enforcement to use 10 threads for every one.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 11 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Ok, I'm removing this part:
```
if (!execSer.awaitTermination(3, TimeUnit.SECONDS)) {
            executorService.shutdown();
        }
```

and just putting this one:
```
executorService.shutdown();
```

p.s. and I renamed to 'executorService' as you're right (yesterday I thought about it but hurried up and didn't do it).

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 12 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @pveentjer Where did you find this piece of code?  There is no such code at all.

```
assertEquals(0, HazelcastClient.getAllHazelcastClients().size());
```

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 13 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @pveentjer
1 - "executorService" was renamed - you're right (I thought about it yesterday too).
2 - before we had "int threadCount=4;" (https://github.com/cherednichenko/hazelcast/commit/108d2c363788624ce4435fbb462b196de4339057#diff-e54d362d7e7f73ac05367fd9784cfe95R73 ) but during discussion/commenting it was removed and rewritten. Should I revert back local 'thread count' ?

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 14 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @pveentjer 
1 - Do you mean at line 69, right ?  Removed.
2 - I'm not sure that I understand - if I remove the threads completely - what would it kept here?

We have just threes tests here:
1:
```@Test
    public void creatingASingleCache() throws URISyntaxException {......} ```

2:
``` @Test
    public void creatingASingleCacheFromMultiProviders() throws URISyntaxException, InterruptedException {}```

3:
``` @Test
    public void creatingMultiCacheFromMultiProviders() throws URISyntaxException, InterruptedException {}```

all those tests for a bit different goals: for single cache, for single + thread count from multi providers (```.getCache("xmlCache");```) and third one for multi cache + + thread count from multi providers (```.getCache("xmlCache"+finalI);```). Honestly the initial idea was from @Danny-Hazelcast and maybe he wants to add something but I said how I understand it - correct me if I'm wrong.

Thanks.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 15 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @pveentjer 
1 - about local variables - answers/discussions is below but this row ```private static final int THREAD_COUNT = 4;``` was added after reviewing/commenting this PR, initially it was local count ```int threadCount=4;``` .
2 - But these tests were created based on real issues which happened/related to .xml configuration that's why actually the tests are using xml configurations. If we rewrote it - it looses initial sense of it, doesn't it? The issue was mentioned at the top of PR (#4753). But if I didn't understand you and what you suggest is within scope of that bug - just give me a bit more information how to do it.

Thanks.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 16 above)
Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: @pveentjer 
1 - removed
2 - Added 'Exception' like: ```public void creatingMultiCacheFromMultiProviders() throws Exception { ```

Thanks.

File: hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java (hunk 12 above)
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: I think I got confused with a different PR.

Comment (cherednichenko) [inline hazelcast/src/test/java/com/hazelcast/cache/CacheCreationTest.java]: Ok.

**Comment by cherednichenko** [comment]
@pveentjer @serkan-ozal Guys - I committed new updates - let's define what should I do with 'thread_local' -- should it be local or global ?

Thanks.

**Comment by serkan-ozal** [comment]
@cherednichenko I am OK with both of two ways (means not a big problem from my side) and choosing one of them should not prevent this PR to be merged.

**Comment by cherednichenko** [comment]
@serkan-ozal I'm also OK with the both versions. 
@pveentjer Just convince me about - should I change to local 'threadCount' variable or keep as is, global one ?

**Comment by pveentjer** [comment]
@cherednichenko keep the constant. Not relevant for now.
